aws/rust-runtime/aws-config/src/imds/client/error.rs

```diff
@@ -0,0 +1,276 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Error types for [`ImdsClient`](crate::imds::client::ImdsClient)
+
+use crate::profile::credentials::ProfileFileError;
+use aws_smithy_client::SdkError;
+use aws_smithy_http::body::SdkBody;
+use http::uri::InvalidUri;
+use std::error::Error;
+use std::fmt;
+
+/// Error context for [`ImdsError::FailedToLoadToken`]
+#[derive(Debug)]
+pub struct FailedToLoadToken {
+    source: SdkError<TokenError>,
+}
+
+impl FailedToLoadToken {
+    /// Returns `true` if a dispatch failure caused the token to fail to load
+    pub fn is_dispatch_failure(&self) -> bool {
+        matches!(self.source, SdkError::DispatchFailure(_))
+    }
+}
+
+/// Error context for [`ImdsError::ErrorResponse`]
+#[derive(Debug)]
+pub struct ErrorResponse {
+    raw: http::Response<SdkBody>,
+}
+
+impl ErrorResponse {
+    /// Returns the raw response from IMDS
+    pub fn response(&self) -> &http::Response<SdkBody> {
+        &self.raw
+    }
+}
+
+/// Error context for [`ImdsError::IoError`]
+#[derive(Debug)]
+pub struct IoError {
+    source: Box<dyn Error + Send + Sync + 'static>,
```

Comment:
We use this so much in so many places in the runtime crates that I wonder if we should type alias this and enshrine it somewhere.